lib/backend/firestore/firestorebk.go

```diff
@@ -541,45 +544,160 @@ func (b *Backend) getRangeDocs(ctx context.Context, startKey, endKey backend.Key
+		}()
+
+		for docSnap, err := range b.documentSnapshots(ctx, params.StartKey.String(), params.EndKey.String(), limit, sort) {
+			if err != nil {
+				yield(backend.Item{}, trace.Wrap(err))
+				return
+			}
+
+			r, err := newRecordFromDoc(docSnap)
+			if err != nil {
+				yield(backend.Item{}, trace.Wrap(err))
+				return
+			}
 
-				if !r.isExpired(b.clock.Now()) {
-					values = append(values, r.backendItem())
+			if r.isExpired(b.clock.Now()) {
+				if _, err := docSnap.Ref.Delete(ctx, firestore.LastUpdateTime(docSnap.UpdateTime)); err != nil && status.Code(err) == codes.FailedPrecondition {
+					// If the document has been updated, then attempt one additional get to see if the
+					// resource was updated and is no longer expired.
+					docSnap, err := b.svc.Collection(b.CollectionName).
+						Doc(docSnap.Ref.ID).
+						Get(ctx)
+					if err != nil {
+						yield(backend.Item{}, trace.Wrap(err))
+						return
+					}
+
+					r, err = newRecordFromDoc(docSnap)
+					if err != nil {
+						yield(backend.Item{}, trace.Wrap(err))
+						return
+					}
+
+					if r.isExpired(b.clock.Now()) {
+						continue
+					}
 				}
 			}
-			// Do not include this document in the results.
-			continue
+
+			if !yield(r.backendItem(), nil) {
+				return
+			}
+			count++
+
+			if limit != backend.NoLimit && count >= limit {
+				return
+			}
+		}
+	}
+}
+
+// documentSnapshots returns an iterator that aggregates all items in the collection
+// in the desired order. This is required because over the years the key has been
+// stored in three formats. In order to iterate over all keys in the collection in the
+// correct order, documents with keys of all formats need to be considered.
+//
+// TODO(tross|tigrato): DELETE IN V19.0.0 with the background migration
+func (b *Backend) documentSnapshots(ctx context.Context, startKey, endKey string, limit int, sort firestore.Direction) iter.Seq2[*firestore.DocumentSnapshot, error] {
+	return func(yield func(*firestore.DocumentSnapshot, error) bool) {
+		snaps := snapshots(b.svc.Collection(b.CollectionName).
+			Where(keyDocProperty, ">=", []byte(startKey)).
+			Where(keyDocProperty, "<=", []byte(endKey)).
+			Limit(limit).
+			OrderBy(keyDocProperty, sort).
+			Documents(ctx))
+
+		legacySnaps := snapshots(b.svc.Collection(b.CollectionName).
+			Where(keyDocProperty, ">=", startKey).
+			Where(keyDocProperty, "<=", endKey).
+			Limit(limit).
+			OrderBy(keyDocProperty, sort).
+			Documents(ctx))
+
+		brokenSnaps := snapshots(b.svc.Collection(b.CollectionName).
+			Where(keyDocProperty, ">=", brokenKey(startKey)).
+			Where(keyDocProperty, "<=", brokenKey(endKey)).
+			Limit(limit).
+			OrderBy(keyDocProperty, sort).
+			Documents(ctx))
+
+		less := func(a, b *firestore.DocumentSnapshot) bool {
+			ar, errA := newRecordFromDoc(a)
+			if errA != nil {
+				return false
+			}
+
+			br, errB := newRecordFromDoc(b)
+			if errB != nil {
+				return true
+			}
+
+			return bytes.Compare(ar.Key, br.Key) < 0
+		}
```

Comment:
`newRecordFromDoc` does a typecheck and then a reflecty unmarshal in the form of `(*firestore.DocumentSnapshot).DataTo`, and this comparison function will get called potentially twice for each document during the stream merging. Map the streams through `newRecordFromDoc` and handle streams of `backend.Item` instead.